SimPEG/potential_fields/gravity/simulation.py

```diff
@@ -27,27 +76,134 @@ class Simulation3DIntegral(BasePFSimulation):
+        # Define jit functions
+        if self.engine == "choclo":
+            if choclo_parallel:
+                self._sensitivity_gravity = _sensitivity_gravity_parallel
+                self._forward_gravity = _forward_gravity_parallel
+            else:
+                self._sensitivity_gravity = _sensitivity_gravity_serial
+                self._forward_gravity = _forward_gravity_serial
+
+    def _sanity_checks_engine(self, kwargs):
+        """
+        Sanity checks for the engine parameter.
+
+        Needs the kwargs passed to the __init__ method to raise some warnings.
+        Will set n_processes to None if it's present in kwargs.
+        """
+        if self.engine not in ("choclo", "geoana"):
+            raise ValueError(
+                f"Invalid engine '{self.engine}'. Choose from 'geoana' or 'choclo'."
+            )
+        if self.engine == "choclo" and choclo is None:
+            raise ImportError(
+                "The choclo package couldn't be found."
+                "Running a gravity simulation with 'engine=\"choclo\"' needs "
+                "choclo to be installed."
+                "\nTry installing choclo with:"
+                "\n    pip install choclo"
+                "\nor:"
+                "\n    conda install choclo"
+            )
+        # Warn if n_processes has been passed
+        if self.engine == "choclo" and "n_processes" in kwargs:
+            warnings.warn(
+                "The 'n_processes' will be ignored when selecting 'choclo' as the "
+                "engine in the gravity simulation.",
+                UserWarning,
+                stacklevel=1,
+            )
+            self.n_processes = None
+        # Sanity checks for sensitivity_path when using choclo and storing in disk
+        if self.engine == "choclo" and self.store_sensitivities == "disk":
+            if os.path.isdir(self.sensitivity_path):
+                raise ValueError(
+                    f"The passed sensitivity_path '{self.sensitivity_path}' is "
+                    "a directory. "
+                    "When using 'choclo' as the engine, 'senstivity_path' "
+                    "should be the path to a new or existing file."
+                )
 
     def fields(self, m):
-        self.model = m
+        """
+        Forward model the gravity field of the mesh on the receivers in the survey
 
+        Parameters
+        ----------
+        m : (n_active_cells,) array
+            Array with values for the model.
+
+        Returns
+        -------
+        (nD,) array
+            Gravity fields generated by the given model on every receiver
+            location.
+        """
+        self.model = m
         if self.store_sensitivities == "forward_only":
             # Compute the linear operation without forming the full dense G
-            fields = mkvc(self.linear_operator())
+            if self.engine == "choclo":
+                fields = self._forward(self.rho)
```

Comment:
Not sure if we want to have different attribute names based on the engine. Feels like the `linear_operator` method should handle the logic...

Reply:
This is related to my reply to the next comment. Since I didn't want to touch the parent class in this PR, I thought having specific private methods for the gravity simulation was the sensitive choice (at least for now).

Also, the design of `linear_operator` + `evaluate_integral` doesn't work well with the new implementation of the simulation: the two for loops (over receivers and over nodes) happen within a single numba-decorated function.

So, considering these two reasons, I think it's better to have these two new private methods in the simulation class: `_forward` and `_sensitivity_matrix`.